### PR TITLE
✨ Add Identity Server to RulesGPT

### PR DESCRIPTION
**Pull request by jackreimers**
Adds authentication via Identity Server, allowing users to log in and receive a JWT that can be used to authenticate with the RulesGPT API.

Related to #98 

There is now a sign in/out button in the top right of the header, when clicked it redirects to SSW Identity Server and upon successful sign in the user is redirected back to RulesGPT.

The token provided from Identity Server can then be used to authenticate with the RulesGPT API. 
Users are not required to sign in and the app will function exactly the same as before.
User sign ins and statistics are not currently traced, this will be added in #102 and #103 

image
**Figure: Using RulesGPT signed out**

image
**Figure: Login screen displayed when sign in is clicked**

image
**Figure: User successfully authenticated and returned to RulesGPT**

### CODE REVIEW

**Comment by jackreimers** [comment]
Cannot be merged until after https://github.com/SSWConsulting/SSW.IdentityServer/pull/93 has been merged.

File: src/Application/Services/OpenAiServiceFactory.cs [hunk 1]
```diff
@@ -1,45 +1,62 @@
...
     }
-    public OpenAIService Create(string? azureOpenAiResourceName)
+
+    public IOpenAIService Create(string? azureOpenAiResourceName)
     {
-        var openAiApiKey = _config["OpenAiApiKey"];
-        var azureOpenAiApiKey = _config["AzureOpenAiApiKey"];
-        var azureOpenAiEndpoint = _config["AzureOpenAiEndpoint"];
-        var useAzureOpenAi = _config.GetValue<bool>("UseAzureOpenAiBool");
+        //TODO: Detect whether user is authenticated
```
Comment (matt-goldman) [inline src/Application/Services/OpenAiServiceFactory.cs:22]: Generally a good practice to leave a link to the PBI for a TODO. Sometimes it's not necessary but in this case as you actually do already have a PBI it couldn't hurt to drop it in here.

Non-blocking comment, just FYI.

File: src/WebAPI/SignalR/RulesHub.cs [hunk 2]
```diff
@@ -66,6 +67,11 @@ public async Task BroadcastMessage(string user, string message)
         CancellationToken cancellationToken
     )
     {
+        if (Context.User.IsAuthenticated())
+        {
+            //TODO: Track user stats
```
Comment (matt-goldman) [inline src/WebAPI/SignalR/RulesHub.cs:72]: Same here - a link to the PBI shows developers that see this code that there is a task allocated for it and that someone is working on it.

In practical terms here that is probably not necessary as you're immediately picking this up and working on it. Definitely worth doing for longer term technical debt though.

**Review by matt-goldman — APPROVED** [review]
Excellent work @jackreimers

File: src/WebAPI/SignalR/RulesHub.cs (hunk 2 above)
Comment (jackreimers) [reply src/WebAPI/SignalR/RulesHub.cs:72]: Noted, thanks for the tip @matt-goldman 🙂

File: src/WebAPI/WebAPI.csproj [hunk 3]
```diff
@@ -4,7 +4,8 @@
         <TargetFramework>net7.0</TargetFramework>
         <Nullable>enable</Nullable>
         <ImplicitUsings>enable</ImplicitUsings>
-        <UserSecretsId>40537d8f-c6e5-4798-980a-3e143e70ad63</UserSecretsId>
+        <!--UserSecretsId>40537d8f-c6e5-4798-980a-3e143e70ad63</UserSecretsId-->
```
Comment (matt-goldman) [inline src/WebAPI/WebAPI.csproj:7]: Remove this comment if it's wrong

**Review by matt-goldman — APPROVED** [review]
LGTM

**Comment by jackreimers** [comment]
@matt-goldman Changes are deployed to staging on RulesGPT and Identity Server and can be tested here: 
https://ashy-meadow-0a2bad900.3.azurestaticapps.net/

It looks to all be working

**Review by MattParkerDev — APPROVED** [review]

**Pull request merged**